### PR TITLE
add dockerfile trt8_cuda11

### PR DESCRIPTION
**Pull request by AlfengYuan**

### CODE REVIEW

**Comment by wang-xinyu** [comment]
@AlfengYuan Thanks!.  Is it better to use NV's official docker directly now?   What's the advantages of this docker?

**Comment by AlfengYuan** [comment]
Of course, it would be better if I could use the official Dockerfile to build my own Docker, but in fact, I failed when I tried to use the official Dockerfile to build my own Docker.

Here is my logs of trying to build docker from the official Dockerfile.

firstly, I download the Dockerfile from [nvidia official tensorrt repository ](https://github.com/NVIDIA/TensorRT/blob/main/docker/ubuntu-20.04.Dockerfile).

secondly, I run `sudo docker build  .` , in where the Dockerfile is, on myself Ubuntu20.04.

Here is the logs:

Sending build context to Docker daemon  8.192kB
Step 1/33 : ARG CUDA_VERSION=11.4.2
Step 2/33 : ARG OS_VERSION=20.04
Step 3/33 : FROM nvidia/cuda:${CUDA_VERSION}-cudnn8-devel-ubuntu${OS_VERSION}
 ---> 9cecbdaa6496
Step 4/33 : LABEL maintainer="NVIDIA CORPORATION"
 ---> Using cache
 ---> e53f5ebba6a4
Step 5/33 : ENV TRT_VERSION 8.2.5.1
 ---> Using cache
 ---> 0e3990a570b1
Step 6/33 : SHELL ["/bin/bash", "-c"]
 ---> Using cache
 ---> 598906ef53c4
Step 7/33 : ARG uid=1000
 ---> Using cache
 ---> e6edc7b3508f
Step 8/33 : ARG gid=1000
 ---> Using cache
 ---> 6ed1f0b8b7db
Step 9/33 : RUN groupadd -r -f -g ${gid} trtuser && useradd -o -r -l -u ${uid} -g ${gid} -ms /bin/bash trtuser
 ---> Using cache
 ---> 4480e1b031b7
Step 10/33 : RUN usermod -aG sudo trtuser
 ---> Using cache
 ---> 50e2717dfbb1
Step 11/33 : RUN echo 'trtuser:nvidia' | chpasswd
 ---> Using cache
 ---> d291d86f151b
Step 12/33 : RUN mkdir -p /workspace && chown trtuser /workspace
 ---> Using cache
 ---> 6e2bd514ed7d
Step 13/33 : ENV DEBIAN_FRONTEND=noninteractive
 ---> Using cache
 ---> 5db4774fccc1
Step 14/33 : RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
 ---> Using cache
 ---> d38b137f7754
Step 15/33 : RUN apt-get update && apt-get install -y software-properties-common
 ---> Using cache
 ---> 2e7d163b6659
Step 16/33 : RUN add-apt-repository ppa:ubuntu-toolchain-r/test
 ---> Using cache
 ---> 42cbbb6f46a7
Step 17/33 : RUN apt-get update && apt-get install -y --no-install-recommends     libcurl4-openssl-dev     wget     zlib1g-dev     git     pkg-config     sudo     ssh     libssl-dev     pbzip2     pv     bzip2     unzip     devscripts     lintian     fakeroot     dh-make     build-essential
 ---> Using cache
 ---> 3c6857f31921
Step 18/33 : RUN apt-get install -y --no-install-recommends       python3       python3-pip       python3-dev       python3-wheel &&    cd /usr/local/bin &&    ln -s /usr/bin/python3 python &&    ln -s /usr/bin/pip3 pip;
 ---> Using cache
 ---> 6c74800a6ee5
Step 19/33 : RUN if [ "${CUDA_VERSION}" = "10.2" ] ; then     v="${TRT_VERSION%.*}-1+cuda${CUDA_VERSION}" &&    apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub &&    apt-get update &&    sudo apt-get install libnvinfer8=${v} libnvonnxparsers8=${v} libnvparsers8=${v} libnvinfer-plugin8=${v}         libnvinfer-dev=${v} libnvonnxparsers-dev=${v} libnvparsers-dev=${v} libnvinfer-plugin-dev=${v}         python3-libnvinfer=${v}; else     v="${TRT_VERSION%.*}-1+cuda${CUDA_VERSION%.*}" &&    apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub &&    apt-get update &&    sudo apt-get install libnvinfer8=${v} libnvonnxparsers8=${v} libnvparsers8=${v} libnvinfer-plugin8=${v}         libnvinfer-dev=${v} libnvonnxparsers-dev=${v} libnvparsers-dev=${v} libnvinfer-plugin-dev=${v}         python3-libnvinfer=${v}; fi
 ---> Using cache
 ---> 9c229950f1db
Step 20/33 : RUN pip3 install --upgrade pip
 ---> Using cache
 ---> bbc03992d497
Step 21/33 : RUN pip3 install setuptools>=41.0.0
 ---> Using cache
 ---> 6c709324a740
Step 22/33 : COPY requirements.txt /tmp/requirements.txt
COPY failed: file not found in build context or excluded by .dockerignore: stat requirements.txt: file does not exist

so finally, I build myself docker image follow the [official install introduce](https://docs.nvidia.com/deeplearning/tensorrt/install-guide/index.html)

My image has the following advantages:
1. Coorect. Make sure it can run our reposity code , I checked on alexnet by python api and arcface by c++ api.
2. Friendly. Just follow my README, every one can have their own tensorrtx docker image,  which contain opencv4.5.5 with dnn module.
3. Network. I have already installed most of the  dependency libraries that you need when you're installing Opencv, Protobuf, or other common libraries.
4. License. This is a docker image I built at home while learning the tensorrtx repository, there will be no license problem.

It's ok, if anyone has better suggestions in the future, anyone is welcome to rp a better tutorial on how to build docker image from the official repository, and how to use docker . but for now, at least the docker I build is working.

**Comment by AlfengYuan** [comment]
at night, I try build from  [official](https://github.com/NVIDIA/TensorRT) again, 
I try use  `./docker/build.sh --file docker/ubuntu-20.04.Dockerfile --tag tensorrt-ubuntu20.04-cuda11.4`

Here is the new logs:
Building container:
> docker build -f docker/ubuntu-20.04.Dockerfile --build-arg CUDA_VERSION=11.4.2 --build-arg uid=1000 --build-arg gid=1000 --tag=tensorrt-ubuntu20.04-cuda11.4 .
Sending build context to Docker daemon  79.36MB
Step 1/33 : ARG CUDA_VERSION=11.4.2
Step 2/33 : ARG OS_VERSION=20.04
Step 3/33 : FROM nvidia/cuda:${CUDA_VERSION}-cudnn8-devel-ubuntu${OS_VERSION}
 ---> 9cecbdaa6496
Step 4/33 : LABEL maintainer="NVIDIA CORPORATION"
 ---> Using cache
 ---> e53f5ebba6a4
Step 5/33 : ENV TRT_VERSION 8.2.5.1
 ---> Using cache
 ---> 0e3990a570b1
Step 6/33 : SHELL ["/bin/bash", "-c"]
 ---> Using cache
 ---> 598906ef53c4
Step 7/33 : ARG uid=1000
 ---> Using cache
 ---> e6edc7b3508f
Step 8/33 : ARG gid=1000
 ---> Using cache
 ---> 6ed1f0b8b7db
Step 9/33 : RUN groupadd -r -f -g ${gid} trtuser && useradd -o -r -l -u ${uid} -g ${gid} -ms /bin/bash trtuser
 ---> Using cache
 ---> 4480e1b031b7
Step 10/33 : RUN usermod -aG sudo trtuser
 ---> Using cache
 ---> 50e2717dfbb1
Step 11/33 : RUN echo 'trtuser:nvidia' | chpasswd
 ---> Using cache
 ---> d291d86f151b
Step 12/33 : RUN mkdir -p /workspace && chown trtuser /workspace
 ---> Using cache
 ---> 6e2bd514ed7d
Step 13/33 : ENV DEBIAN_FRONTEND=noninteractive
 ---> Using cache
 ---> 5db4774fccc1
Step 14/33 : RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
 ---> Using cache
 ---> d38b137f7754
Step 15/33 : RUN apt-get update && apt-get install -y software-properties-common
 ---> Using cache
 ---> 2e7d163b6659
Step 16/33 : RUN add-apt-repository ppa:ubuntu-toolchain-r/test
 ---> Using cache
 ---> 42cbbb6f46a7
Step 17/33 : RUN apt-get update && apt-get install -y --no-install-recommends     libcurl4-openssl-dev     wget     zlib1g-dev     git     pkg-config     sudo     ssh     libssl-dev     pbzip2     pv     bzip2     unzip     devscripts     lintian     fakeroot     dh-make     build-essential
 ---> Using cache
 ---> 3c6857f31921
Step 18/33 : RUN apt-get install -y --no-install-recommends       python3       python3-pip       python3-dev       python3-wheel &&    cd /usr/local/bin &&    ln -s /usr/bin/python3 python &&    ln -s /usr/bin/pip3 pip;
 ---> Using cache
 ---> 6c74800a6ee5
Step 19/33 : RUN if [ "${CUDA_VERSION}" = "10.2" ] ; then     v="${TRT_VERSION%.*}-1+cuda${CUDA_VERSION}" &&    apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub &&    apt-get update &&    sudo apt-get install libnvinfer8=${v} libnvonnxparsers8=${v} libnvparsers8=${v} libnvinfer-plugin8=${v}         libnvinfer-dev=${v} libnvonnxparsers-dev=${v} libnvparsers-dev=${v} libnvinfer-plugin-dev=${v}         python3-libnvinfer=${v}; else     v="${TRT_VERSION%.*}-1+cuda${CUDA_VERSION%.*}" &&    apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub &&    apt-get update &&    sudo apt-get install libnvinfer8=${v} libnvonnxparsers8=${v} libnvparsers8=${v} libnvinfer-plugin8=${v}         libnvinfer-dev=${v} libnvonnxparsers-dev=${v} libnvparsers-dev=${v} libnvinfer-plugin-dev=${v}         python3-libnvinfer=${v}; fi
 ---> Using cache
 ---> 9c229950f1db
Step 20/33 : RUN pip3 install --upgrade pip
 ---> Using cache
 ---> bbc03992d497
Step 21/33 : RUN pip3 install setuptools>=41.0.0
 ---> Using cache
 ---> 6c709324a740
Step 22/33 : COPY requirements.txt /tmp/requirements.txt
 ---> Using cache
 ---> a8d321f161df
Step 23/33 : RUN pip3 install -r /tmp/requirements.txt
 ---> Running in 34e9703f0a41
Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com
Looking in links: https://download.pytorch.org/whl/cu113/torch_stable.html
Collecting onnx==1.10.2
  Downloading onnx-1.10.2-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (12.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 58.9 kB/s eta 0:00:00
Collecting tensorflow-gpu==2.5.1
  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/d9/29/cf71bbd2dd98a31c4e054d505406c84c465704a2bffba85dd2b467537e0a/tensorflow_gpu-2.5.1-cp38-cp38-manylinux2010_x86_64.whl
  Downloading tensorflow_gpu-2.5.1-cp38-cp38-manylinux2010_x86_64.whl (454.5 MB)


this time it blocked in step23 for a long time and report ReadTimeoutError when downloading required whl package. I'm not sure it will be succeed in the end, I guess it's probably because of the domestic network. 

In conclusion, I personally prefer to use my own docker image.  That's why I came up with this RP.

**Comment by wang-xinyu** [comment]
@AlfengYuan Thanks,  got it,   maybe it's fetching tf from google,  so it's blocked by GFW.
Merging it.
